x-pack/plugins/apm/public/components/shared/EnvironmentFilter/index.tsx

```diff
@@ -0,0 +1,123 @@
+// @ts-ignore
+import { EuiSuperSelect, EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import React, { useState, useEffect } from 'react';
+import styled from 'styled-components';
+import { fontSizes, units, px } from '../../../style/variables';
+import { useFetcher } from '../../../hooks/useFetcher';
+import { loadServiceEnvironments } from '../../../services/rest/apm/services';
+import { useLocation } from '../../../hooks/useLocation';
+import { useUiFilters } from '../../../hooks/useUiFilters';
+import { useUrlParams } from '../../../hooks/useUrlParams';
+import { history } from '../../../utils/history';
+import { fromQuery, toQuery } from '../Links/url_helpers';
+
+const anyEnvironmentValue = 'All';
+const nullEnvironmentValue = 'Null';
+
+function updateEnvironmentUrl(
+  location: ReturnType<typeof useLocation>,
+  environment?: string
+) {
+  const nextEnvironmentQueryParam =
+    environment !== anyEnvironmentValue ? environment : undefined;
+  history.push({
+    ...location,
+    search: fromQuery({
+      ...toQuery(location.search),
+      environment: nextEnvironmentQueryParam
+    })
+  });
+}
+
+const OptionInputContainer = styled.div`
+  min-width: 128px;
+`;
+
+const EnvironmentFilterLabel = styled.div`
+  height: 100%;
+  border-width: 1px 0 1px 1px;
+  border-style: solid;
+  border-color: #e1e1e1;
+  background: #eee;
+  font-size: ${fontSizes.small};
+  padding: ${px(units.minus)};
+`;
+
+export const EnvironmentFilter: React.FC = () => {
+  const location = useLocation();
```

Comment:
_Following is not really related to this PR but I felt like writing it somewhere :D_

I found out that we can simply get `location` from `history.location`. This is normally not recommended since `history.location` is mutable, and thus cannot be used to determine whether location has changed. But in our case we don't want to compare location: we just use `location` to construct a url.

Disadvantages with `useLocation` (this also applies to `useUrlParams`):
Whenever the location changes, all components that subscribe to location changes (consume `useLocation`) will re-render. This can be costly if the component is complex.

I made a PR that applies this thinking to a few components: https://github.com/elastic/kibana/pull/35939

cc @jasonrhodes

Reply:
I wonder if it's clearer and less dangerous to just use `useMemo` somehow with `useLocation` so that it's clear we are just asking for the value once and don't care if it changes, rather than reaching into `history` and forgetting when that's okay and when it's not?

Reply:
That's an interesting thought, and probably worth investigating.
My first thought is that hooks must be called unconditionally, so I don't think we can simply wrap it in `useMemo` - but there might be another way to achieve that.